# Day 1 Reading Journal

Hi everyone.  These are the exercises for the first set of readings from ThinkPython.  While this journal consists of several required exercises, it is meant to encourage active reading more generally.  You should use this journal to take detailed notes, catalog questions, and explore the content from ThinkPython deeply.

For more information on using the Jupyter notebook, check out the documentation [here](https://www.datacamp.com/community/tutorials/tutorial-jupyter-notebook##UseJupyter).

Additionally, in order to write text in this document, you will be making use of Markdown cells. A handy cheatsheet for Markdown is available [here](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet).

## Chapter 1

Note: the exercise numbers below match up with the reading for ease of cross referencing.

### Exercise 1.4

Start the Python interpreter and use it as a simple calculator. Python’s syntax for math operations is almost the same as standard mathematical notation. For example, the symbols +, - and / denote addition, subtraction and division, as you would expect. The symbol for multiplication is *

* How many seconds are there in 42 minutes 42 seconds?
* How many miles are there in 10 kilometers? Hint: there are 1.61 kilometers in a mile.
* If you run a 10 kilometer race in 42 minutes 42 seconds, what is your average pace (time per mile in minutes and seconds)? What is your average speed in miles per hour?

Please use the following markdown cell to provide your answers.

I am not sure exactly how you wanted me to enter my answers. I am currently using the cells to show my code, but it is possible that you only wanted the answers. I will try to ask a ninja, but in the meanwhile I will be using the cells to show my code.

In [3]:
42 * 60 + 42

2562

In [2]:
10 / 1.61

6.211180124223602

In [4]:
2562 / (10/1.61)

412.482

In [11]:
412.482//60

6.0

In [6]:
412.482 % 60

52.48200000000003

In [7]:
(10/1.61)/(2562/3600)

8.727653570337614

There are 2562 seconds in 42 minutes and 42 seconds

10 km is 6.21 miles

Your average pace is 6 minutes and 52.48 seconds per mile

Your average speed is 8.73 mph.

## Chapter 2

### Exercise

Assume that we execute the following assignment statements:

```
width = 17
height = 12.0
delimiter = '.'
```

For each of the following expressions, write the value of the expression and the type (of the value of the expression).

Example: `width * 2` <br/>
Answer: 34, integer

1. `width/2`
2. `width/2.0`
3. `height/3`
4. `1 + 2 * 5`
5. `delimiter * 5`

Please use the following markdown cell to provide your answer.

width = 17
height = 12.0
delimiter = '.'

1) 8.5, float (in other python2 it would be 8, int, but python3 division automatically converts to float unless you specify //)

2) 8.5, float

3) 4.0, float

4) 11, int

5) '.....', string


### Exercise 2.2

Practice using the Python interpreter as a calculator.  For each of these provide Python code that computes each answer.

The volume of a sphere with radius r is $4/3 \pi  r^3$. What is the volume of a sphere with radius 5? <br />
Hint: 392.7 is wrong! You can approximate pi with 3.14 or import it through the Python math library.

In [16]:
import math
r = 5
(4/3) * math.pi * math.pow(r,3)

523.5987755982989

Suppose the cover price of a book is \$24.95, but bookstores get a 40% discount. Shipping costs \$3 for the first copy and 75 cents for each additional copy. What is the total wholesale cost for 60 copies?

In [17]:
price = 24.95
discount = .6
first_ship = 3
all_but_first_ship = .75
books = 60

price * discount * books + first_ship + all_but_first_ship * .75

901.7624999999999

If I leave my house at 6:52 am and run 1 mile at an easy pace (8:15 per mile), then 3 miles at tempo (7:12 per mile) and 1 mile at easy pace again, what time do I get home for breakfast? 

In [28]:
easy = 8 * 60 + 15
tempo = 7 * 60 + 12
t_sec = 2 * easy + 3 * tempo

hours = t_sec // 3600
mins = (t_sec // 60) % 60
secs = t_sec % 60

hour = str(6 + hours + ((52 + mins) // 60))
minute = str((52 + mins) % 60)
second = str(secs)

if secs > 10:
    print (hour + ':' + minute + ':' + second)
else:
    print (hour + ':' + minute + ':' + '0' + second)

7:30:06


## Chapter 3

### Exercise 3.1

Write a function named `right_justify` that takes a string named s as a parameter and prints the string with enough leading spaces so that the last letter of the string is in column 70 of the display.

```
>>> right_justify('monty')
                                                                 monty
```
Hint: Use string concatenation and repetition. Also, Python provides a built-in function called `len` that returns the length of a string, so the value of `len('monty')` is 5.

In [35]:
def right_justify (s):
    "this function takes a string and right justifies it to column 70"
    s = str(s)
    length = len(s)
    leading = 70 - length
    for x in range(leading):
        s = ' ' + s
    print(s)
    return


right_justify('monty')

                                                                 monty


### Exercise 3.3

Note: This exercise should be done using only the statements and other features we have learned so far.

1\. Write a function that draws a grid like the following:

```
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
```
Note: There are four empty spaces between the vertical bars.

To print more than one value on a line, you can print a comma-separated sequence of values:

```
print('+', '-')
```

By default, `print` advances to the next line, but you can override that behavior and put a space at the end, like this:

```
print('+', end=' ')
print('-')
```

The output of these statements is '+ -'.

A `print` statement with no argument ends the current line and goes to the next line.

In [42]:
def grid2x2 ():
    "Prints a grid. function is designed to be easily scalable."
    boxlen = 5
    gridsize = 2
    gridlen = boxlen * gridsize + 1
    for y in range(gridlen):
        for x in range(gridlen):
            if y % boxlen == 0:
                if x % boxlen == 0:
                    print('+', end=' ')
                else:
                    print('-', end=' ')
            else:
                if x % boxlen == 0:
                    print('|', end=' ')
                else:
                    print(' ', end=' ')
        print('')
    return()
            
                    
grid2x2()

+ - - - - + - - - - + 
|         |         | 
|         |         | 
|         |         | 
|         |         | 
+ - - - - + - - - - + 
|         |         | 
|         |         | 
|         |         | 
|         |         | 
+ - - - - + - - - - + 


()

2\. Write a function that draws a similar grid with four rows and four columns.

In [46]:
def grid4x4 ():
    "Prints a grid. function is designed to be easily scalable."
    boxlen = 5
    gridsize = 4
    gridlen = boxlen * gridsize + 1
    for y in range(gridlen):
        for x in range(gridlen):
            if y % boxlen == 0:
                if x % boxlen == 0:
                    print('+', end=' ')
                else:
                    print('-', end=' ')
            else:
                if x % boxlen == 0:
                    print('|', end=' ')
                else:
                    print(' ', end=' ')
        print('')
    return
            
                    
grid4x4()

+ - - - - + - - - - + - - - - + - - - - + 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
+ - - - - + - - - - + - - - - + - - - - + 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
+ - - - - + - - - - + - - - - + - - - - + 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
+ - - - - + - - - - + - - - - + - - - - + 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
+ - - - - + - - - - + - - - - + - - - - + 


## Chapter 5.1 - 5.7

### Exercise 5.2

Fermat’s Last Theorem says that there are no positive integers $a$, $b$, and $c$ such that $$a^n + b^n = c^n$$ for any values of $n$ greater than 2.

1\. Write a function named `check_fermat` that takes four parameters – $a$, $b$, $c$ and $n$ – and that checks to see if Fermat’s theorem holds. If n is greater than 2 and it turns out to be true that
$a^n + b^n = c^n$, then the program should print: "Holy smokes, Fermat was wrong!" Otherwise the program should print: "No, that doesn’t work."

In [ ]:
def check_fermat(a,b,c,n):
    "checks that fermat's last theorem holds for the input"
    if (math.pow(a,n) + math.pow(b,n)) == math.pow(c,n) and n > 2:
        print("Holy smokes, Fermat was wrong!")
    else:
        print("No, that doesn't work.")
    return

2\. Write a function that prompts the user to input values for $a$, $b$, $c$ and $n$, converts them to integers, and uses `check_fermat` to check whether they violate Fermat’s theorem.

In [51]:
def check_fermat(a,b,c,n):
    "checks that fermat's last theorem holds for the input"
    if (math.pow(a,n) + math.pow(b,n)) == math.pow(c,n) and n > 2:
        print("Holy smokes, Fermat was wrong!")
    else:
        print("No, that doesn't work.")
    return

def get_fermat_test():
    a = int(input ("Please enter an integer to be used as 'a'."))
    b = int(input ("Please enter an integer to be used as 'b'."))
    c = int(input ("Please enter an integer to be used as 'c'."))
    n = int(input ("Please enter an integer to be used as 'n'."))
    check_fermat(a,b,c,n)
    return

get_fermat_test()

Please enter an integer to be used as 'a'.4
Please enter an integer to be used as 'b'.4
Please enter an integer to be used as 'c'.4
Please enter an integer to be used as 'n'.4
No, that doesn't work.


### Exercise 5.3 

If you are given three sticks, you may or may not be able to arrange them in a triangle. For example, if one of the sticks is 12 inches long and the other two are one inch long, it is clear that you will not be able to get the short sticks to meet in the middle. For any three lengths, there is a simple test to see if it is possible to form a triangle:
> If any of the three lengths is greater than the sum of the other two, then you cannot form a triangle. Otherwise, you can. (If the sum of two lengths equals the third, they form what is called a “degenerate” triangle.)

1\. Write a function named `is_triangle` that takes three integers as arguments, and that prints either "Yes" or "No," depending on whether you can or cannot form a triangle from sticks with the given lengths.

In [2]:
def is_triangle(a,b,c):
    if a > b + c or b > a + c or c > a + b:
        print("No")
    else:
        print("Yes")
    return


2\. Write a function that prompts the user to input three stick lengths, converts them to integers, and uses `is_triangle` to check whether sticks with the given lengths can form a triangle.

In [3]:
def is_triangle(a,b,c):
    "Checks if three lengths can form a triangle"
    if a > b + c or b > a + c or c > a + b:
        print("No")
    else:
        print("Yes")
    return


def get_lengths():
    "Asks for length of 3 sticks, then checks if they can form a triangle"
    a = int(input("Please enter the length of stick 1:"))
    b = int(input("Please enter the length of stick 2:"))
    c = int(input("Please enter the length of stick 3:"))
    
    is_triangle(a,b,c)
    return


get_lengths()

Please enter the length of stick 1:1
Please enter the length of stick 2:1
Please enter the length of stick 3:1
Yes


## Quick poll
About how long did you spend working on this Reading Journal?

Aproximately 2 hours.

## Notes for the Instructors

Please use the space below to write comments to help us plan the next class session.  For instance, if you want to see us go over an example of a particular type of problem, you can indicate that here.

Please remember that the best way to get quick feedback from the instructors as well as your peers, is to use Piazza.  However, consider using this channel if it doesn't make sense to post your note to Piazza.